pyomo/repn/util.py

```diff
@@ -221,6 +236,195 @@ def __rpow__(self, other):
+            else:
+                self[child_type] = self._before_invalid
+        elif not hasattr(child, 'is_expression_type'):
+            if check_if_numeric_type(child):
+                self[child_type] = self._before_native
+            else:
+                self[child_type] = self._before_invalid
+        elif not child.is_expression_type():
+            if child.is_potentially_variable():
+                self[child_type] = self._before_var
+            else:
+                self[child_type] = self._before_param
+        elif not child.is_potentially_variable():
+            self[child_type] = self._before_npv
+            pv_base_type = child.potentially_variable_base_class()
+            if pv_base_type not in self:
+                try:
+                    child.__class__ = pv_base_type
+                    self.register_dispatcher(visitor, child)
+                finally:
+                    child.__class__ = child_type
+        elif (
+            issubclass(child_type, _named_subexpression_types)
+            or child_type is kernel.expression.noclone
+        ):
+            self[child_type] = self._before_named_expression
+        else:
+            self[child_type] = self._before_general_expression
+        return self[child_type](visitor, child)
+
+    @staticmethod
+    def _before_general_expression(visitor, child):
+        return True, None
+
+    @staticmethod
+    def _before_native(visitor, child):
+        return False, (_CONSTANT, child)
+
+    @staticmethod
+    def _before_complex(visitor, child):
+        return False, (_CONSTANT, complex_number_error(child, visitor, child))
+
+    @staticmethod
+    def _before_invalid(visitor, child):
+        return False, (
+            _CONSTANT,
+            InvalidNumber(
+                child, f"{child!r} ({type(child)}) is not a valid numeric type"
+            ),
+        )
+
+    @staticmethod
+    def _before_string(visitor, child):
+        return False, (
+            _CONSTANT,
+            InvalidNumber(
+                child, f"{child!r} ({type(child)}) is not a valid numeric type"
+            ),
+        )
+
+    @staticmethod
+    def _before_npv(visitor, child):
+        try:
+            return False, (
+                _CONSTANT,
+                visitor.handle_constant(visitor.evaluate(child), child),
+            )
+        except (ValueError, ArithmeticError):
+            return True, None
+
+    @staticmethod
+    def _before_param(visitor, child):
+        return False, (_CONSTANT, visitor.handle_constant(child.value, child))
+
+    #
+    # The following methods must be defined by derivative classes (along
+    # with any other special-case handling they want to implement;
+    # usually including handling for Monomial, Linear, and
+    # ExternalFunction
+    #
+
+    # @staticmethod
+    # def _before_var(visitor, child):
+    #     pass
+
+    # @staticmethod
+    # def _before_named_expression(visitor, child):
+    #     pass
```

Comment:
Why are these commented out?

Reply:
Mostly because I don't like ABCMeta. These are two methods that are required by `register_dispatcher`, but are not implemented in the base class (the LP and NL implementations are completely different).  They are commented out to remind developers of an expected API...

Reply:
But isn't the advantage of ABCMeta that it will actually yell at you when you don't implement the things you expect to be there?